concordium-consensus/tests/e2e/EndToEndTests/CredentialDeploymentTests.hs

```diff
@@ -0,0 +1,328 @@
+import Data.FileEmbed
+import qualified Data.Vector as Vec
+import Lens.Micro.Platform
+import Test.HUnit
+import Test.Hspec
+
+import Concordium.Common.Version
+import Concordium.GlobalState.BlockState
+import Concordium.ID.Types
+import Concordium.KonsensusV1.TestMonad
+import Concordium.KonsensusV1.TreeState.Implementation
+import Concordium.KonsensusV1.TreeState.Types
+import Concordium.KonsensusV1.Types
+import Concordium.Types
+import Concordium.Types.HashableTo
+import Concordium.Types.Option
+import Concordium.Types.Transactions
+import EndToEndTests.E2ETestData
+
+-- | Helper for reading an 'AccountCreation' from a 'ByteString'.
+readAccountCreation :: BSL.ByteString -> AccountCreation
+readAccountCreation bs =
+    case AE.eitherDecode bs of
+        Left err -> error $ "Cannot read account creation " ++ err
+        Right d -> if vVersion d == 0 then vValue d else error "Incorrect account creation version."
+
+-- 3 valid credentials
+{-# WARNING cred1 "Do not use in production." #-}
+cred1 :: AccountCreation
+cred1 = readAccountCreation . BSL.fromStrict $ $(makeRelativeToProject "testdata/initial-credential-1.json" >>= embedFile)
+
+{-# WARNING cred2 "Do not use in production." #-}
+cred2 :: AccountCreation
+cred2 = readAccountCreation . BSL.fromStrict $ $(makeRelativeToProject "testdata/initial-credential-2.json" >>= embedFile)
+
+{-# WARNING cred3 "Do not use in production." #-}
+cred3 :: AccountCreation
+cred3 = readAccountCreation . BSL.fromStrict $ $(makeRelativeToProject "testdata/credential-1.json" >>= embedFile)
+
+-- | A credential deployment transaction yielding cred1.
+credBi1 :: BlockItem
+credBi1 =
+    credentialDeployment $ addMetadata (\x -> CredentialDeployment{biCred = x}) (tt + 1) cred1
+  where
+    tt = utcTimeToTransactionTime testTime
+
+-- | A credential deployment transaction yielding cred2.
+credBi2 :: BlockItem
+credBi2 =
+    credentialDeployment $ addMetadata (\x -> CredentialDeployment{biCred = x}) (tt + 1) cred2
+  where
+    tt = utcTimeToTransactionTime testTime
+
+-- | A credential deployment transaction yielding cred3
+credBi3 :: BlockItem
+credBi3 =
+    credentialDeployment $ addMetadata (\x -> CredentialDeployment{biCred = x}) (tt + 1) cred3
+  where
+    tt = utcTimeToTransactionTime testTime
+
+-- | Valid block for round 1 with 1 credential deployment.
+testBB1 :: BakedBlock
+testBB1 =
+    BakedBlock
+        { bbRound = 1,
+          bbEpoch = 0,
+          bbTimestamp = 1_000,
+          bbBaker = bakerId,
+          bbQuorumCertificate = genesisQuorumCertificate genesisHash,
+          bbTimeoutCertificate = Absent,
+          bbEpochFinalizationEntry = Absent,
+          bbNonce = computeBlockNonce genesisLEN 1 (bakerVRFKey bakerId),
+          bbTransactions = Vec.fromList [credBi1],
+          bbTransactionOutcomesHash = read "b9444648bf759471276fdba1930af0c543847d22de89c27939791898d757516d",
+          bbStateHash = read "b8bc96ec5f162db36784ea96ec29e3e8ad92abff341a6847e3bf524fdada28ff"
```

Comment:
Please explain where these come from. How do we maintain this?

Reply:
The `TestMonad` has a logger instance which will log errors. If any of the hashes does not match then the test will fail and the execution environment will log it. So one can grab the hashes from there. 

These will need to change when e.g. the hashing scheme is changed.

Reply:
Ok, but my point is that if somebody is to maintain this in the future, what are they supposed to do?

It's not clear, so please document this. Nobody will find this comment in a year when this needs to be updated. Explain in the code what the rationale is.